File: backend/modules/photonlang/ops/registry.py
# backend/modules/photonlang/ops/registry.py
from dataclasses import dataclass
from pathlib import Path
import json
from typing import Dict, List, Literal, Optional

@dataclass
class OpLaws:
    commutative: bool = False
    associative: bool = False
    idempotent: bool = False

@dataclass
class OpSpec:
    glyph: str
    name: str
    arity: Literal["1","2","n"]
    precedence: int
    associativity: Literal["left","right","none"]
    laws: OpLaws
    lowering_call: str
    hover: Optional[str] = None
    examples: Optional[List[str]] = None

def load_specs() -> Dict[str, OpSpec]:
    p = Path(__file__).with_name("operators.json")
    data = json.loads(p.read_text(encoding="utf-8"))
    out: Dict[str,OpSpec] = {}
    for row in data["operators"]:
        laws = OpLaws(**row.get("laws", {}))
        spec = OpSpec(
            glyph=row["glyph"],
            name=row["name"],
            arity=row["arity"],
            precedence=row.get("precedence", 50),
            associativity=row.get("associativity","none"),
            laws=laws,
            lowering_call=row.get("lowering",{}).get("call", row["name"]),
            hover=row.get("hover"),
            examples=row.get("examples", []),
        )
        out[spec.glyph] = spec
    return out

File: tools/photon-js/src/registry.ts
import fs from "node:fs";
import path from "node:path";

export type OperatorSpec = { glyph: string; name: string; arity: string; };
export type Registry = Record<string, OperatorSpec>;

export function loadRegistry(): Registry {
  const p = process.env.PHOTON_OPS_JSON ??
    path.resolve(process.cwd(), "backend/modules/photonlang/ops/operators.json");
  const raw = JSON.parse(fs.readFileSync(p, "utf8"));
  const out: Registry = {};
  for (const op of raw.operators as OperatorSpec[]) out[op.glyph] = op;
  return out;
}